{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "da667d01-72f3-4924-b8a1-798bc6439594",
   "metadata": {},
   "source": [
    "# Wastyk et al. (2021) from the Sonnenburg Lab showed in a feeding study the inter-individual differences in inflammatory immune response after intervention with a high-fiber diet. Here, we use microbial sequencing data from this study to compare the results of MICOM metabolic modeling with the results obtained by the authors. Specifically, we are curious whether the SCFA production profiles from our metabolic models agree with the inflammatory immune response group used by the authors to categorize study participants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4511dabc-3b38-4a3c-bbd8-be671bf1307f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyreadr\n",
    "import pandas as pd\n",
    "import micom\n",
    "# import qiime2 as q2\n",
    "from plotnine import *\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "\n",
    "import os\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3ef1eb9-c761-46bd-9f85-d58bc95fd82e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Get to our directory and pull in a dictionary with the inflammation groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "681eae9f-e571-4946-8d96-96097cdd40b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{8007: 'HighInflammation',\n",
       " 8009: 'HighInflammation',\n",
       " 8013: 'HighInflammation',\n",
       " 8001: 'HighInflammation',\n",
       " 8002: 'HighInflammation',\n",
       " 8023: 'LowInflammationI',\n",
       " 8029: 'LowInflammationI',\n",
       " 8017: 'LowInflammationI',\n",
       " 8041: 'LowInflammationI',\n",
       " 8022: 'LowInflammationI',\n",
       " 8003: 'LowInflammationI',\n",
       " 8006: 'LowInflammationI',\n",
       " 8039: 'LowInflammationII',\n",
       " 8018: 'LowInflammationII',\n",
       " 8036: 'LowInflammationII',\n",
       " 8037: 'LowInflammationII',\n",
       " 8038: 'LowInflammationII',\n",
       " 8035: 'LowInflammationII'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir('/proj/gibbons/nbohmann/exvivo/sonnenburg_2021/')\n",
    "groups = (pd.read_csv('fiber_immune_groups.csv',index_col = 1)['Immune_group'].str.replace('group1','HighInflammation')\n",
    "          .str.replace('group2','LowInflammationI').str.replace('group3','LowInflammationII').astype('str').to_dict())\n",
    "groups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7edf591a-1962-41bd-86f7-7db10b71e5f3",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Pull in their abundance tables, and filter to start and end timepoints. Convert table to use in MICOM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "40094b94-74cf-4f78-8448-6abc364e376c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV_1</th>\n",
       "      <th>ASV_10</th>\n",
       "      <th>ASV_100</th>\n",
       "      <th>ASV_101</th>\n",
       "      <th>ASV_102</th>\n",
       "      <th>ASV_103</th>\n",
       "      <th>ASV_104</th>\n",
       "      <th>ASV_105</th>\n",
       "      <th>ASV_106</th>\n",
       "      <th>ASV_107</th>\n",
       "      <th>...</th>\n",
       "      <th>ASV_90</th>\n",
       "      <th>ASV_91</th>\n",
       "      <th>ASV_92</th>\n",
       "      <th>ASV_93</th>\n",
       "      <th>ASV_94</th>\n",
       "      <th>ASV_95</th>\n",
       "      <th>ASV_96</th>\n",
       "      <th>ASV_97</th>\n",
       "      <th>ASV_98</th>\n",
       "      <th>ASV_99</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8001_1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.060921</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003467</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8001_2</th>\n",
       "      <td>0.000424</td>\n",
       "      <td>0.000562</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.033276</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000944</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000012</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8001_3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.116271</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005932</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8001_4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.091430</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002735</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8001_5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.081136</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002343</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8041_5</th>\n",
       "      <td>0.002461</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8041_6</th>\n",
       "      <td>0.003770</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8041_7</th>\n",
       "      <td>0.001856</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8041_8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8041_9</th>\n",
       "      <td>0.003360</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>162 rows × 503 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ASV_1    ASV_10  ASV_100  ASV_101   ASV_102   ASV_103  ASV_104  \\\n",
       "sample_id                                                                      \n",
       "8001_1     0.000000  0.000000      0.0      0.0  0.060921  0.000000      0.0   \n",
       "8001_2     0.000424  0.000562      0.0      0.0  0.033276  0.000131      0.0   \n",
       "8001_3     0.000000  0.000000      0.0      0.0  0.116271  0.000000      0.0   \n",
       "8001_4     0.000000  0.000000      0.0      0.0  0.091430  0.000000      0.0   \n",
       "8001_5     0.000000  0.000000      0.0      0.0  0.081136  0.000000      0.0   \n",
       "...             ...       ...      ...      ...       ...       ...      ...   \n",
       "8041_5     0.002461  0.000000      0.0      0.0  0.000000  0.000000      0.0   \n",
       "8041_6     0.003770  0.000000      0.0      0.0  0.000000  0.000000      0.0   \n",
       "8041_7     0.001856  0.000000      0.0      0.0  0.000000  0.000000      0.0   \n",
       "8041_8     0.000000  0.000000      0.0      0.0  0.000000  0.000000      0.0   \n",
       "8041_9     0.003360  0.000000      0.0      0.0  0.000000  0.000000      0.0   \n",
       "\n",
       "            ASV_105   ASV_106   ASV_107  ...  ASV_90    ASV_91  ASV_92  \\\n",
       "sample_id                                ...                             \n",
       "8001_1     0.000000  0.000000  0.003467  ...     0.0  0.000000     0.0   \n",
       "8001_2     0.000944  0.000018  0.000000  ...     0.0  0.000367     0.0   \n",
       "8001_3     0.000000  0.000000  0.005932  ...     0.0  0.000000     0.0   \n",
       "8001_4     0.000000  0.000000  0.002735  ...     0.0  0.000000     0.0   \n",
       "8001_5     0.000000  0.000000  0.002343  ...     0.0  0.000000     0.0   \n",
       "...             ...       ...       ...  ...     ...       ...     ...   \n",
       "8041_5     0.000000  0.000000  0.000000  ...     0.0  0.000000     0.0   \n",
       "8041_6     0.000000  0.000000  0.000000  ...     0.0  0.000000     0.0   \n",
       "8041_7     0.000000  0.000000  0.000000  ...     0.0  0.000000     0.0   \n",
       "8041_8     0.000000  0.000000  0.000000  ...     0.0  0.000000     0.0   \n",
       "8041_9     0.000000  0.000000  0.000000  ...     0.0  0.000000     0.0   \n",
       "\n",
       "           ASV_93    ASV_94    ASV_95  ASV_96  ASV_97  ASV_98    ASV_99  \n",
       "sample_id                                                                \n",
       "8001_1        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "8001_2        0.0  0.000012  0.000208     0.0     0.0     0.0  0.000249  \n",
       "8001_3        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "8001_4        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "8001_5        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "...           ...       ...       ...     ...     ...     ...       ...  \n",
       "8041_5        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "8041_6        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "8041_7        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "8041_8        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "8041_9        0.0  0.000000  0.000000     0.0     0.0     0.0  0.000000  \n",
       "\n",
       "[162 rows x 503 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir('/proj/gibbons/nbohmann/exvivo/sonnenburg_2021/')\n",
    "abundance = pyreadr.read_r('tip_glom_counts.rds')[None]\n",
    "abundance = abundance[(abundance['Participant'].isin(groups.keys()))] #only use individuals in the fiber study\n",
    "abundance['Participant'] = abundance['Participant'].astype('str') \n",
    "abundance['sample_id'] = abundance['Participant']+'_'+abundance['Timepoint'] #we'll include timepoint in the index so we don't lose that info\n",
    "abundance.set_index('sample_id',inplace = True,drop = True)\n",
    "abundance = abundance.drop(columns = ['Participant','Group','Group_value','Timepoint']) #need only ASV counts\n",
    "abundance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "865ab731-700b-457f-aa7e-c0d8b9509c47",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Also pull in their taxonomy table, assinging reads to taxa. We'll use this to genus depth, and convert to use in MICOM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "987720e1-a3f2-4c98-908c-559019105e3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>id</th>\n",
       "      <th>abundance</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8001_1</td>\n",
       "      <td>Archaea;Euryarchaeota;Methanobacteria;Methanob...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Methanobrevibacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8001_1</td>\n",
       "      <td>Archaea;Euryarchaeota;Methanobacteria;Methanob...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Methanosphaera</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8001_1</td>\n",
       "      <td>Archaea;Euryarchaeota;Thermoplasmata;E2;[metha...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Methanomassiliicoccus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8001_1</td>\n",
       "      <td>Archaea;Euryarchaeota;Thermoplasmata;E2;[metha...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Vadinca11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8001_1</td>\n",
       "      <td>Bacteria;Actinobacteria;Actinobacteria;Actinom...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Actinomyces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15223</th>\n",
       "      <td>8041_9</td>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;En...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Escherichia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15224</th>\n",
       "      <td>8041_9</td>\n",
       "      <td>Bacteria;Proteobacteria;Gammaproteobacteria;En...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Proteus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15225</th>\n",
       "      <td>8041_9</td>\n",
       "      <td>Bacteria;Spirochaetes;[brachyspirae];[brachysp...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Brachyspira</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15226</th>\n",
       "      <td>8041_9</td>\n",
       "      <td>Bacteria;Synergistetes;Synergistia;Synergistal...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Pyramidobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15227</th>\n",
       "      <td>8041_9</td>\n",
       "      <td>Bacteria;Verrucomicrobia;Verrucomicrobiae;Verr...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Akkermansia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15228 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id                                                 id  abundance  \\\n",
       "0        8001_1  Archaea;Euryarchaeota;Methanobacteria;Methanob...        0.0   \n",
       "1        8001_1  Archaea;Euryarchaeota;Methanobacteria;Methanob...        0.0   \n",
       "2        8001_1  Archaea;Euryarchaeota;Thermoplasmata;E2;[metha...        0.0   \n",
       "3        8001_1  Archaea;Euryarchaeota;Thermoplasmata;E2;[metha...        0.0   \n",
       "4        8001_1  Bacteria;Actinobacteria;Actinobacteria;Actinom...        0.0   \n",
       "...         ...                                                ...        ...   \n",
       "15223    8041_9  Bacteria;Proteobacteria;Gammaproteobacteria;En...        0.0   \n",
       "15224    8041_9  Bacteria;Proteobacteria;Gammaproteobacteria;En...        0.0   \n",
       "15225    8041_9  Bacteria;Spirochaetes;[brachyspirae];[brachysp...        0.0   \n",
       "15226    8041_9  Bacteria;Synergistetes;Synergistia;Synergistal...        0.0   \n",
       "15227    8041_9  Bacteria;Verrucomicrobia;Verrucomicrobiae;Verr...        0.0   \n",
       "\n",
       "                       genus  \n",
       "0         Methanobrevibacter  \n",
       "1             Methanosphaera  \n",
       "2      Methanomassiliicoccus  \n",
       "3                  Vadinca11  \n",
       "4                Actinomyces  \n",
       "...                      ...  \n",
       "15223            Escherichia  \n",
       "15224                Proteus  \n",
       "15225            Brachyspira  \n",
       "15226         Pyramidobacter  \n",
       "15227            Akkermansia  \n",
       "\n",
       "[15228 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxa = pyreadr.read_r('rep_ASV_labels.rds')[None].set_index('rep_ASV_label').drop(columns = ['rep_ASV','Species'])\n",
    "taxa.index.rename('Feature ID',inplace = True) \n",
    "for col in taxa.columns:  #convert so it can be read by AGORA \n",
    "    taxa[col] = taxa[col].str.split('_').str[-1]\n",
    "taxa['Genus'].replace('', np.nan, inplace = True)\n",
    "taxa.dropna(how = 'any',inplace = True)\n",
    "taxa = taxa.apply(lambda row: \";\".join(row.str.capitalize().fillna(\"\")), axis=1).to_frame() #into MICOM accepted format\n",
    "taxa.columns = [\"Taxon\"]\n",
    "abundance = abundance[taxa.index] #we only want ASVs in the abundance table that are present in the taxonomy table\n",
    "taxa = taxa['Taxon'].to_dict()\n",
    "abundance = abundance.rename(columns=taxa).reset_index()\n",
    "abundance = pd.melt(abundance, id_vars = 'sample_id', value_vars = abundance.columns[:-1],\n",
    "                    var_name = 'id', value_name = 'abundance') # melt into long form df \n",
    "abundance = abundance.groupby(by = ['sample_id','id']).sum().reset_index()\n",
    "abundance['genus'] = abundance['id'].str.split(';').str[-1] # need a genus column in df \n",
    "abundance"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb5fe044-19bd-4ef1-9e48-27efdc88b4e7",
   "metadata": {},
   "source": [
    "## Before building our models we need to define our model database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "22f73bb4-9e85-4284-af67-956ae951f223",
   "metadata": {},
   "outputs": [],
   "source": [
    "agora = ('/proj/gibbons/refs/micom_dbs/agora103_genus.qza')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e94e8c0f-6b40-4c5b-8447-0c12dd488055",
   "metadata": {},
   "source": [
    "## Now we'll build the models using our abundance table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "183f2adc-78db-4e25-8baf-63363e7ad3ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "models = mm.workflows.build(abundance,out_folder = '/proj/gibbons/nbohmann/exvivo/sonnenburg_2021/models/',\n",
    "                      model_db = agora, cutoff = 0.001, threads = 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e6d65cc-6849-4e7e-a96d-8184dc66b496",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Let's try growing all samples at all timepoints on a high fiber diet, and seeing if there is a trend in SCFA production during intervention. First we'll look at the fiber consumption profile, and scale the diets by that value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "11a12596-daed-4875-8a62-baa065800f3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9776ddeda87a488a82d6aa766d04f9dd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"></pre>\n"
      ],
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/users/nbohmann/.local/lib/python3.8/site-packages/micom/workflows/grow.py:199: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n"
     ]
    }
   ],
   "source": [
    "os.chdir('/proj/gibbons/nbohmann/exvivo/sonnenburg_2021/')\n",
    "manifest = pd.read_csv('models/data/manifest.csv')\n",
    "os.chdir('/proj/gibbons/nbohmann/exvivo/sonnenburg_2021/')\n",
    "hf_medium = pd.read_csv('/proj/gibbons/nbohmann/exvivo/diets/highfiber.tsv',sep = '\\t').rename(\n",
    "    columns = {'Reaction':'reaction','Flux Value':'flux'})\n",
    "growth_hf = micom.workflows.grow(manifest, model_folder='models/data',\n",
    "                                 medium = hf_medium, tradeoff = 0.7, strategy = 'none',threads = 10,presolve = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "182407b7-6a47-470b-ba0f-5accc8fdada4",
   "metadata": {},
   "source": [
    "## Now we'll collect all the exchanges resulting from the growth of our models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b7c2dd9-091b-4274-9d58-a61378e626d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "exchanges = growth_hf.exchanges\n",
    "exchanges['timepoint'] = exchanges['sample_id'].str.split('_').str[1]\n",
    "exchanges['sample_id'] = exchanges['sample_id'].str.split('_').str[0]\n",
    "exchanges = (exchanges[exchanges.direction == \"export\"].groupby( #get total production\n",
    "    [\"sample_id\",\"metabolite\", \"reaction\",\"timepoint\"]).apply(lambda df:  \n",
    "    sum(df.flux * df.abundance)).reset_index())\n",
    "exchanges"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3825b050-b0e9-4b31-af93-cf2703a801f0",
   "metadata": {},
   "source": [
    "## Build a dataframe with all the SCFA predictions, removing timepoints 8 and 9 as these were after the study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebd39993-d1be-417a-9586-d9c84525c7b9",
   "metadata": {},
   "outputs": [],
   "source": [
    " scfa = (exchanges[\n",
    "    (exchanges['reaction'].str.startswith('EX_but(e)'))|\n",
    "    (exchanges['reaction'].str.startswith('EX_ppa(e)'))|\n",
    "    (exchanges['reaction'].str.startswith('EX_ac(e)'))]\n",
    "         .rename(columns = {0:'flux'}))\n",
    "scfa['timepoint'] = scfa['timepoint'].astype('str')\n",
    "scfa['group'] = scfa['sample_id'].astype('int').map(groups)\n",
    "scfa = pd.pivot_table(scfa,index = ['sample_id','timepoint'], columns = 'metabolite', values = 'flux').reset_index()\n",
    "scfa['total'] = scfa['but[e]']+scfa['ppa[e]']\n",
    "scfa['group'] = scfa['sample_id'].astype('int').map(groups)\n",
    "scfa = scfa[~(scfa['timepoint'].str.contains('8'))& ~(scfa['timepoint'].str.contains('9'))]\n",
    "scfa['timepoint'] = scfa['timepoint'].astype('int')\n",
    "scfa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e1e9026-edb7-44ae-aa4c-bd4f6404ae56",
   "metadata": {},
   "source": [
    "## Now, plot the predcited SCFA production among each inflammation group, at each timepoint across the intervention and average. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4464de4-0075-480f-8484-780715ed0f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt1 = (ggplot(scfa,\n",
    "    aes(x = 'timepoint', y = 'ac[e]'))\n",
    "    +geom_point(color = 'cadetblue')\n",
    "    +geom_smooth(method = 'lm')\n",
    "    +labs(x = 'Timepoint', y = 'Predicted Acetate \\n ($\\dfrac{mmol}{gDW*h}$)')\n",
    "    +facet_wrap(facets = 'group')\n",
    "    +theme(text = element_text(size=20),panel_background=element_rect(fill = \"white\",\n",
    "                                colour = \"white\",size = 0.5, linetype = \"solid\"),panel_grid= element_blank(),\n",
    "                                axis_line = element_line(size = 2, linetype = \"solid\",colour = \"black\"),\n",
    "                                axis_line_y = element_blank(), axis_text_x=element_text(rotation = 20, hjust = 1)))\n",
    "plt1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d5bb5b0-3ea4-48e6-b76b-c115f427e2a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt2 = (ggplot(scfa,\n",
    "    aes(x = 'timepoint', y = 'total'))\n",
    "    +geom_point(color = 'cadetblue')\n",
    "    +geom_smooth(method = 'lm')\n",
    "    +labs(x = 'Timepoint', y = 'Predicted Butyrate+Propionate \\n ($\\dfrac{mmol}{gDW*h}$)')\n",
    "    +facet_wrap(facets = 'group')\n",
    "    +theme(text = element_text(size=20),panel_background=element_rect(fill = \"white\",\n",
    "                                colour = \"white\",size = 0.5, linetype = \"solid\"),panel_grid= element_blank(),\n",
    "                                axis_line = element_line(size = 2, linetype = \"solid\",colour = \"black\"),\n",
    "                                axis_line_y = element_blank(), axis_text_x=element_text(rotation = 20, hjust = 1)))\n",
    "plt2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6d2ed4c-e662-447e-bc46-183b1c833b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt3 = (ggplot(\n",
    "        scfa[(scfa.timepoint!='8')&(scfa.timepoint!='9')], aes(x = 'group', y = 'ac[e]'))\n",
    "        +geom_boxplot(fill = 'cadetblue',size = 1, width = 0.4)\n",
    "        +labs(x = 'Immune Response Group', y = 'Predicted Acetate ($\\dfrac{mmol}{L*h}$)')\n",
    "        +scale_x_discrete(limits = ['HighInflammation','LowInflammationI','LowInflammationII'],\n",
    "                          labels = ['High \\n Inf.', 'Low \\n Inf. I', 'Low \\n Inf. II'])\n",
    "        +theme(text = element_text(size=20, color = 'black'),panel_background=element_rect(fill = \"white\",\n",
    "                                colour = \"white\",size = 0.5, linetype = \"solid\"),panel_grid= element_blank(),\n",
    "                                axis_line = element_line(size = 2, linetype = \"solid\",colour = \"black\"),\n",
    "                                legend_position='right',axis_text_x=element_text(rotation = 20, hjust = 1)))\n",
    "plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c44a19d5-d549-40f0-b417-8ce1466066bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt4 = (ggplot(\n",
    "        scfa[(scfa.timepoint!='8')&(scfa.timepoint!='9')], aes(x = 'group', y = 'total[e]'))\n",
    "        +geom_boxplot(fill = 'cadetblue',size = 1, width = 0.4)\n",
    "        +labs(x = 'Immune Response Group', y = 'Predicted Butyrate + Propionate($\\dfrac{mmol}{L*h}$)')\n",
    "        +scale_x_discrete(limits = ['HighInflammation','LowInflammationI','LowInflammationII'],\n",
    "                          labels = ['High \\n Inf.', 'Low \\n Inf. I', 'Low \\n Inf. II'])\n",
    "        +theme(text = element_text(size=20, color = 'black'),panel_background=element_rect(fill = \"white\",\n",
    "                                colour = \"white\",size = 0.5, linetype = \"solid\"),panel_grid= element_blank(),\n",
    "                                axis_line = element_line(size = 2, linetype = \"solid\",colour = \"black\"),\n",
    "                                legend_position='right',axis_text_x=element_text(rotation = 20, hjust = 1)))\n",
    "plt4"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
